src/main/java/servlets/OAuthCallbackServlet.java

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package servlets;
+
+import com.google.gson.JsonParser;
+import java.io.IOException;
+import java.io.FileInputStream;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse.BodyHandlers;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonElement;
+
+/** A servlet to request access and refresh tokens. */
+@WebServlet("/api/oauth/callback/sheets")
+public class OAuthCallbackServlet extends HttpServlet {
+  // Token request parameters.
+  private final String CLIENT_ID = "client_id=150737768611-svndjtlklolq53g4ass4r3sqal2i31p5.apps.googleusercontent.com";
+  private final String GRANT_TYPE = "grant_type=authorization_code";
+  private final String CODE = "https://accounts.google.com/o/oauth2/v2/auth";
+  private final String TOKEN_URI = "https://oauth2.googleapis.com/token";
+
+  // Figure out a way to secure method to retrieve the client secret.
+  private final String CLIENT_SECRET = "";
+  private final String REDIRECT_URI = "redirect_uri= SOME CALLBACK URI";
+
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String error = request.getParameter("error");
+    String code = request.getParameter("code");
```

Comment:
What do the two different strings called code refer to? Can you make their differences more clear?

Reply:
+1